google-cloud-firestore/lib/google/cloud/firestore/batch.rb

```diff
@@ -0,0 +1,1020 @@
+      #
+      # All changes are accumulated in memory until the block passed to
+      # {Database#batch} completes. Unlike transactions, batches are not
+      # automatically retried. See {Database#batch}.
+      #
+      # Batched writes have fewer failure cases than transactions and use
+      # simpler code. They are not affected by contention issues, because they
+      # don't depend on consistently reading any documents.
+      #
+      # @see https://firebase.google.com/docs/firestore/manage-data/transactions
+      #   Transactions and Batched Writes
+      #
+      # @example
+      #   require "google/cloud/firestore"
+      #
+      #   firestore = Google::Cloud::Firestore.new
+      #
+      #   firestore.batch do |b|
+      #     # Set the data for NYC
+      #     b.col("cities").doc("NYC").set({ name: "New York City" })
+      #
+      #     # Update the population for SF
+      #     b.col("cities").doc("SF").update({ population: 1000000 })
+      #
+      #     # Delete LA
+      #     b.col("cities").doc("LA").delete
+      #   end
+      #
+      class Batch
+        ##
+        # New Batch object.
+        def initialize
+          @writes = []
+        end
+
+        ##
+        # The project resource the Cloud Firestore batch belongs to.
+        #
+        # @return [Project] project resource
+        def project
+          @database.project
+        end
+
+        ##
+        # The database resource the Cloud Firestore batch belongs to.
+        #
+        # @return [Database] database resource.
+        def database
+          @database
+        end
+
+        ##
+        # The project identifier for the Cloud Firestore database.
+        #
+        # @return [String] project identifier.
+        def project_id
+          @database.project_id
+        end
+
+        ##
+        # The database identifier for the Cloud Firestore batch.
+        #
+        # @return [String] database identifier.
+        def database_id
+          @database.database_id
+        end
+
+        ##
+        # The full Database path for the Cloud Firestore batch.
+        #
+        # @return [String] database resource path.
+        def path
+          @database.path
+        end
+
+        ##
+        # Retrieves a list of collections.
+        #
+        # @yield [collections] The block for accessing the collections.
+        # @yieldparam [Collection::Reference] collection A collection.
+        #
+        # @return [Enumerator<Collection::Reference>] collection list.
+        #
+        # @example
+        #   require "google/cloud/firestore"
+        #
+        #   firestore = Google::Cloud::Firestore.new
+        #
+        #   firestore.batch do |b|
+        #     # Get the root collections
+        #     b.cols.each do |col|
+        #       # Print the collection
+        #       puts col.collection_id
+        #     end
+        #   end
+        #
+        def cols
```

Comment:
Yes, I think this is something that will be popular with Rubyists. This is needed to support an additional style coding for using batches. For example, the simplest way to update a couple of documents is as follows:

```ruby
# Update multiple documents, makes multiple API calls
firestore.col("cities").doc("NYC").update({ name: "Big Apple" })
firestore.col("cities").doc("SF").update({ name: "San Fran" })
```

That isn't very efficient, as it makes multiple API calls. Instead, users can change how they perform the updates to make a single API call by using a batch:

```ruby
# create document references
nyc_ref = firestore.doc "cities/NYC"
sf_ref = firestore.doc "cities/SF"

# Update multiple document in the same batch
firestore.batch do |b|
  b.update(nyc_ref, { name: "Big Apple" })
  b.update(sf_ref, { name: "San Fran" })
end
```

While this code is more performant than the first example, it isn't as direct. Rubyists have a strong aesthetic for "beautiful code". So while this PR definitely supports this style, we also want to support the following:

```ruby
# Update multiple document in batch, makes a single API call
firestore.batch do |b|
  b.col("cities").doc("NYC").update({ name: "Big Apple" })
  b.col("cities").doc("SF").update({ name: "San Fran" })
end
```

By allowing a Batch to act like the other context we can support this style of coding. Collections and documents that were created in a Batch will add their updates to the Batch as long as the Batch is open. I think this will be a big benefit for Rubyists by allowing them the ability to structure their code this way.